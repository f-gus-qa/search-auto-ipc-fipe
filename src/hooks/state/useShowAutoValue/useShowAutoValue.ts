import _ from 'lodash';
import useSelectState from '../useSelectState';

function useShowAutoValue() {
  const selectedAuto = useSelectState();
  const {
    brand, model, year, value,
  } = selectedAuto;

  const isDone = !Object.values(selectedAuto).includes(null);
  const autoName = isDone
    ? [brand, model, year].map(_.curry(_.get)(_, 'name')).join(' ').trim()
    : null;

  return { isDone, autoName, value };
}

export default useShowAutoValue;
