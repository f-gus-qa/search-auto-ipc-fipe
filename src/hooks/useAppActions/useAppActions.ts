import { actions } from '@/actions';
import useAppContext from '../useAppContext';

type Actions = typeof actions
type ActionKey = keyof Actions
type DispatchActions = { [K in ActionKey]: (...args: Parameters<Actions[K]>) => void }

function useAppActions() {
  const { dispatch } = useAppContext();

  return Object.keys(actions).reduce((allActions, actionKey) => {
    const action = actions[actionKey as ActionKey];
    return {
      ...allActions,
      [actionKey]: (...args: Parameters<typeof action>) => (
        // @ts-ignore: Unreachable code error
        dispatch(action(...args))
      ),
    };
  }, {}) as DispatchActions;
}

export default useAppActions;
