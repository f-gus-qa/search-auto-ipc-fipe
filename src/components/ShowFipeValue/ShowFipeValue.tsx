/* eslint-disable jsx-a11y/anchor-is-valid */
import { Link, makeStyles, Typography } from '@material-ui/core';
import { useEffect, useState } from 'react';
import useShowAutoValue from '@/hooks/state/useShowAutoValue';
import useAppActions from '@/hooks/useAppActions';

const useStyles = makeStyles(({ spacing, palette }) => ({
  containerFipeValue: {
    textAlign: 'center',
    padding: `${spacing(8)}px 0px`,
    display: 'grid',
    gridRowGap: '22px',
  },
  valueLabel: {
    color: 'white',
    padding: '8px 24px',
    borderRadius: '36px',
    backgroundColor: palette.success.main,
  },
}));

function ShowFipeValue() {
  const classes = useStyles();
  const autoValues = useShowAutoValue();
  const [{ autoName, value }, setDelayValues] = useState(autoValues);
  const { resetState } = useAppActions();

  useEffect(() => {
    if (autoValues.isDone) {
      setDelayValues(autoValues);
    }
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [autoValues.isDone]);

  const handleReset = () => {
    resetState();
  };

  return (
    <div className={classes.containerFipeValue}>
      <div>
        <Typography variant="h4">
          Tabela Fipe:
          {' '}
          {autoName}
        </Typography>
      </div>
      <div>
        <Typography component="span" className={classes.valueLabel} variant="h4">
          {value}
        </Typography>
      </div>
      <div>
        <Typography component="div" variant="subtitle2">
          Este é o preço de compra do veículo
        </Typography>
        <Link
          data-testid="linkReset"
          onClick={handleReset}
          component="button"
          variant="body2"
        >
          Escolher outro veículo
        </Link>
      </div>
    </div>
  );
}

export default ShowFipeValue;
