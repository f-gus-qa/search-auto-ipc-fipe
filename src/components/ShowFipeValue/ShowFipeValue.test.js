import React from 'react';
import {
  render,
  screen,
  fireEvent,
} from '@testing-library/react';
import {
  act as rendererAct,
  create,
} from 'react-test-renderer';
import useShowAutoValue from '@/hooks/state/useShowAutoValue';
import useAppActions from '@/hooks/useAppActions';
import ShowFipeValue from '.';
import {
  fixtureAutoValueComplete,
  fixtureAutoValueIncomplete,
} from '@/fixture/autoValue';

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useEffect: jest.fn(),
}));
jest.mock('@/hooks/state/useShowAutoValue', () => jest.fn());
jest.mock('@/hooks/useAppActions', () => jest.fn());

const component = <ShowFipeValue />;
const resetState = jest.fn();
const autoValueComplete = fixtureAutoValueComplete();
const autoValueIncomplete = fixtureAutoValueIncomplete();
useAppActions.mockReturnValue({ resetState });

describe('<ShowFipeValue>', () => {
  describe('when search is concluded', () => {
    beforeEach(() => {
      useShowAutoValue.mockReturnValue(autoValueComplete);
    });

    it('check renders', () => {
      rendererAct(() => {});
      const { toJSON } = create(component);

      expect(toJSON()).toMatchSnapshot();
    });

    it('check showing correct values', () => {
      render(component);

      expect(screen.getByText(
        `Tabela Fipe: ${autoValueComplete.autoName}`,
      )).toBeInTheDocument();

      expect(screen.getByText(
        autoValueComplete.value,
      )).toBeInTheDocument();
    });

    it('checks reset link', () => {
      render(component);

      const link = screen.getByTestId('linkReset');
      fireEvent.click(link);

      expect(resetState).toHaveBeenCalledTimes(1);
    });
  });

  describe('when search not is concluded', () => {
    beforeEach(() => {
      useShowAutoValue.mockReturnValue(autoValueIncomplete);
    });

    it('check renders', () => {
      rendererAct(() => {});
      const { toJSON } = create(component);

      expect(toJSON()).toMatchSnapshot();
    });
  });
});
