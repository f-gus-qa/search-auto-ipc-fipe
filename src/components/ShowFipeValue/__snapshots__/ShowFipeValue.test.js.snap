// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ShowFipeValue> when search is concluded check renders 1`] = `
<div
  className="makeStyles-containerFipeValue-1"
>
  <div>
    <h4
      className="MuiTypography-root MuiTypography-h4"
    >
      Tabela Fipe:
       
      some-name
    </h4>
  </div>
  <div>
    <span
      className="MuiTypography-root makeStyles-valueLabel-2 MuiTypography-h4"
    >
      some-value
    </span>
  </div>
  <div>
    <div
      className="MuiTypography-root MuiTypography-subtitle2"
    >
      Este é o preço de compra do veículo
    </div>
    <button
      className="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiLink-button MuiTypography-body2 MuiTypography-colorPrimary"
      data-testid="linkReset"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
    >
      Escolher outro veículo
    </button>
  </div>
</div>
`;

exports[`<ShowFipeValue> when search not is concluded check renders 1`] = `
<div
  className="makeStyles-containerFipeValue-1"
>
  <div>
    <h4
      className="MuiTypography-root MuiTypography-h4"
    >
      Tabela Fipe:
       
    </h4>
  </div>
  <div>
    <span
      className="MuiTypography-root makeStyles-valueLabel-2 MuiTypography-h4"
    />
  </div>
  <div>
    <div
      className="MuiTypography-root MuiTypography-subtitle2"
    >
      Este é o preço de compra do veículo
    </div>
    <button
      className="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiLink-button MuiTypography-body2 MuiTypography-colorPrimary"
      data-testid="linkReset"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
    >
      Escolher outro veículo
    </button>
  </div>
</div>
`;
